waddrmgr/scoped_manager.go

```diff
@@ -777,55 +780,60 @@ func (s *ScopedKeyManager) nextAddresses(ns walletdb.ReadWriteBucket,
 				info.branch, info.index, adtChain,
 			)
 			if err != nil {
-				return nil, maybeConvertDbError(err)
+				return nil, nil, maybeConvertDbError(err)
 			}
 		case *scriptAddress:
 			encryptedHash, err := s.rootManager.cryptoKeyPub.Encrypt(a.AddrHash())
 			if err != nil {
 				str := fmt.Sprintf("failed to encrypt script hash %x",
 					a.AddrHash())
-				return nil, managerError(ErrCrypto, str, err)
+				return nil, nil, managerError(ErrCrypto, str, err)
 			}
 
 			err = putScriptAddress(
 				ns, &s.scope, a.AddrHash(), ImportedAddrAccount,
 				ssNone, encryptedHash, a.scriptEncrypted,
 			)
 			if err != nil {
-				return nil, maybeConvertDbError(err)
+				return nil, nil, maybeConvertDbError(err)
 			}
 		}
 	}
 
-	// Finally update the next address tracking and add the addresses to
-	// the cache after the newly generated addresses have been successfully
-	// added to the db.
 	managedAddresses := make([]ManagedAddress, 0, len(addressInfo))
 	for _, info := range addressInfo {
 		ma := info.managedAddr
-		s.addrs[addrKey(ma.Address().ScriptAddress())] = ma
-
-		// Add the new managed address to the list of addresses that
-		// need their private keys derived when the address manager is
-		// next unlocked.
-		if s.rootManager.IsLocked() && !s.rootManager.WatchOnly() {
-			s.deriveOnUnlock = append(s.deriveOnUnlock, info)
-		}
-
 		managedAddresses = append(managedAddresses, ma)
 	}
 
-	// Set the last address and next address for tracking.
-	ma := addressInfo[len(addressInfo)-1].managedAddr
-	if internal {
-		acctInfo.nextInternalIndex = nextIndex
-		acctInfo.lastInternalAddr = ma
-	} else {
-		acctInfo.nextExternalIndex = nextIndex
-		acctInfo.lastExternalAddr = ma
+	// Finally return a closure that will update the next address tracking
+	// and add the addresses to the cache after the newly generated
+	// addresses have been successfully added to the db.
+	onCommit := func() {
```

Comment:
We'll need to be careful with this, as we need to ensure that all access to the caches are properly thread-safe. The callback MUST be executed right at the point of the caller.

Reply:
That's true... Seems for this to be thread safe, mutexes must be added around the db access + `onCommit` calls. 

Wondering if there's a reason to keep the current style of passing in the db transaction/bucket from the outer level, or if we instead should try changing to opening the db+updating the caches only at the point where the data is put?